### DISCUSSION TITLE
What's the usage of lastSubmission parameter in useForm?

### DISCUSSION THREAD

**Discussion by anderson-chong** [edmundhung/conform#353] [Q&A]
This is a simplified example, but I have a Zod schema that not only used for validation, but change the key of the object properties after parsing, mainly because the API requires a different set of keys.

```typescript
const Schema = z
  .object({ name: z.string() })
  .transform(({ name }) => ({ username: name }));
```

Hence, I was wondering what is the consequences of this when it's used by Conform? And I realised the only time Remix's `actionData.submission` was used (which is the only place that contains the parsed `formData`) is in the `lastSubmission` parameter of `useForm`.

```typescript
  const [form, { selection, startAt, endAt, vendors }] = useForm({
    ...
    constraint: getFieldsetConstraint(Schema),
    lastSubmission: actionData?.submission,
    onValidate: ({ formData }) => parse(formData, { schema: Schema }),
    ...
  });
```

But what's the usage of `lastSubmission` parameter? And is there any potential consequences I'm not aware of?

**Comment by edmundhung** [comment]
> Hence, I was wondering what is the consequences of this when it's used by Conform

This is a valid pattern. You should be able to get the transformed value from `submission.value`.

> But what's the usage of lastSubmission parameter? And is there any potential consequences I'm not aware of?

There are two main purpose of `lastSubmission`:
1. To get the latest error from the server after form submission.
2. Progressive enhancement. `lastSubmission` has all kind of information required to persists the form value and allows it to work consistently even when the full document is reloaded.

However, the parsed value (i.e. `submission.value`) is actually redundant and could be deleted before sending the submission back. There will be an RFC tomorrow with new APIs that improve how submission is handled and returned to the client.
